src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs

```diff
@@ -19,255 +18,99 @@ namespace Microsoft.CodeAnalysis.PatternMatching
     /// Also, while the pattern matcher is culture aware, it uses the culture specified in the
     /// constructor.
     /// </summary>
-    internal sealed partial class PatternMatcher : IDisposable
+    internal abstract partial class PatternMatcher : IDisposable
     {
+        private static readonly char[] s_dotCharacterArray = { '.' };
+
         public const int NoBonus = 0;
         public const int CamelCaseContiguousBonus = 1;
         public const int CamelCaseMatchesFromStartBonus = 2;
         public const int CamelCaseMaxWeight = CamelCaseContiguousBonus + CamelCaseMatchesFromStartBonus;
 
-        private static readonly char[] s_dotCharacterArray = { '.' };
-
         private readonly object _gate = new object();
 
+        private readonly bool _includeMatchedSpans;
         private readonly bool _allowFuzzyMatching;
-        private readonly bool _invalidPattern;
-        private readonly PatternSegment _fullPatternSegment;
-        private readonly PatternSegment[] _dotSeparatedPatternSegments;
 
         private readonly Dictionary<string, StringBreaks> _stringToWordSpans = new Dictionary<string, StringBreaks>();
         private static readonly Func<string, StringBreaks> _breakIntoWordSpans = StringBreaker.BreakIntoWordParts;
 
         // PERF: Cache the culture's compareInfo to avoid the overhead of asking for them repeatedly in inner loops
         private readonly CompareInfo _compareInfo;
 
+        private bool _invalidPattern;
```

Comment:
I would prefer to not take a virtual method hit inside a method that is already virtual.  The purpose of the matcher is to try to be low overhead.  The normal case for it is to be called against hundreds of thousands of strings.